onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -359,6 +363,25 @@ Status CUDAExecutionProvider::SetComputeStream(void* stream) {
   return Status::OK();
 }
 
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11000
+void CUDAExecutionProvider::CaptureBegin()  {
+  ORT_THROW_IF_ERROR(Sync());
+  graph_.CaptureBegin();
+}
+
+void CUDAExecutionProvider::CaptureEnd() {
+  graph_.CaptureEnd();
+  ORT_THROW_IF_ERROR(Sync());
+}
+
+void CUDAExecutionProvider::Replay() {
+  ORT_THROW_IF_ERROR(OnRunStart());
```

Comment:
Since graph replaying is not thread-safe, do we need to call Replay() within a critical section here ?

Reply:
yes, it is needed. I will add a lock here.

Reply:
Added now.